Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs

```diff
@@ -47,18 +52,23 @@ internal sealed class BlockCaret : IBlockCaret
             }
         }
 
+        private static readonly Point s_invalidPoint = new Point(double.NaN, double.NaN);
+
+        private readonly IVimHost _vimHost;
         private readonly ITextView _textView;
         private readonly IProtectedOperations _protectedOperations;
         private readonly IEditorFormatMap _editorFormatMap;
         private readonly IClassificationFormatMap _classificationFormatMap;
         private readonly IAdornmentLayer _adornmentLayer;
-        private readonly object _tag = new object();
+        private readonly List<object> _tags = new List<object>();
+        private readonly HashSet<object> _adornmentsPresent = new HashSet<object>();
         private readonly DispatcherTimer _blinkTimer;
         private readonly IControlCharUtil _controlCharUtil;
-        private CaretData? _caretData;
+
+        private List<VirtualSnapshotPoint> _caretPoints = new List<VirtualSnapshotPoint>();
+        private Dictionary<int, CaretData> _caretDataMap = new Dictionary<int, CaretData>();

```

Comment:
Rather than have four different collections that need to be kept in sync for each caret I think it would be easier to have a single collection and push all the data into a new type (or possibly augment `CaretData`) to have the new type. Consider if we had this type:

```
internal sealed class CaretViewData {
  internal CaretData { get; set; }
  internal bool IsAdornmentPresent { get; set; }
  internal object Tag { get; } = new object();
  internal VirtualSnapshotPoint Point { get; set; }
}
```

Seems like that could allow us to move to a single collection to manager here and would make some of the code below a bit simpler.

Reply:
I like this idea. By way of explanation for doing the way that I did, because the block caret is so central to Vim, I tried to make the logic exactly parallel the old block caret, especially for the primary caret but also for all carets. I was pleasantly surprised that the multi-caret support "just worked" and I didn't have any trouble with it. I'll take a look at streamlining it.

Reply:
I've looked long and hard at this idea, and in my opinion, the data structures exactly match the expected behavior and performance characteristics of their respective operations. For example, the tag array does and should have a different cardinality than caret data map. The adornments present map cannot be converted to an array field without maintaining a separate "double index" to allow efficient lookup. The caret points themselves are simply a cached snapshot of an external interface and doesn't imply that **any** carets are actually visible. Finally, the caret data map leverages the immutability of caret data in its values.